src/python/zquantum/core/estimator.py

```diff
@@ -84,6 +84,67 @@ def get_context_selection_circuit_for_group(
     return context_selection_circuit, transformed_operator
 
 
+def allocate_shots(
+    shot_allocation_strategy: str,
+    frame_operators: List[IsingOperator],
+    n_samples: Optional[int] = None,
+    n_total_samples: Optional[int] = None,
+    prior_expectation_values: Optional[ExpectationValues] = None,
+) -> List[int]:
+    """Generates the number of shots for each frame operator, using either the "uniform"
+    or the "optimal" shot allocation.
+
+    Args:
+        shot_allocation_strategy: Which shot allocation strategy to use. "uniform" attributes the same
+                                  number of shots to each frame operator, whereas "optimal" divides the
+                                  shots among all frame_operators according to their coefficients and
+                                  variances if prior expectation values are provided.
+        frame_operators: The list of IsingOperators that will be evaluated
+        n_samples: The number of samples per frame operator using the uniform allocation strategy
+                   Exactly one of n_samples or n_total_samples must be provided
+        n_total_samples: The total number of samples across all frame operators when using the optimal
+                         allocation strategy.
+                         Exactly one of n_samples or n_total_samples must be provided
+        prior_expectation_values: If given, this is used to estimate the variances of the frame operators when
+                                  doing optimal shot allocation.
+    Returns:
+        List of integers giving the number of shots for each frame operator.
+    """
+    if shot_allocation_strategy not in (
+        "optimal",
+        "uniform",
+    ):
+        raise ValueError(
```

Comment:
Another way of doing it would be to put this raise in the`else` statement after `if shot_allocation_strategy == "uniform":` and `if shot_allocation_strategy == "optimal":` .
Though I'm not sure if that's a definitely better solution, just different, so it can stay the way it is :)